crates/wasmparser/src/validator/component.rs

```diff
@@ -1776,19 +1783,50 @@ impl ComponentState {
         }
 
         let ty = self.defined_type_at(ty, offset)?;
-        let ComponentDefinedType::Future(_) = &types[ty] else {
+        let ComponentDefinedType::Future(elem_ty) = &types[ty] else {
             bail!(offset, "`future.write` requires a future type")
         };
 
-        let ty_id = self
-            .check_options(types, &options, offset)?
-            .require_memory(offset)?
-            .check_lower(offset)?
-            .check_core_type(
-                types,
-                FuncType::new([ValType::I32; 2], [ValType::I32]),
-                offset,
-            )?;
+        let options = self.check_options(types, &options, offset)?;
+        options.check_lower(offset)?;
+
+        let LoweredFuncType::New(func_ty) = ComponentFuncType {
+            info: TypeInfo::new(),
+            params: if let Some(elem_ty) = elem_ty {
+                Box::new([(KebabString::new("p").unwrap(), *elem_ty)])
+            } else {
+                Box::new([])
+            },
+            result: None,
+        }
+        .lower(
+            types,
+            &CanonicalOptions {
+                string_encoding: StringEncoding::Utf8,
+                memory: options.memory,
+                realloc: None,
+                post_return: None,
+                concurrency: Concurrency::Async { callback: None },
+                core_type: None,
+                gc: false,
+            },
```

Comment:
Instead of creating a new `CanonicalOptions` could this `clone()` and manipulate the `options` above? I mostly want to keep things in sync in terms of gc/string encoding/etc for future compatibliity

Reply:
I was thinking the same thing, but then check with Luke, who confirmed that we use `MAX_FLAT_ASYNC_PARAMS` for `future.write` _regardless_ of whether the `future.write` is called sync or async.  I guess I could still clone, but then force the `concurrence` to async.

Reply:
The other wrinkle is that `ComponentFuncType::lower` may return a `LoweredFuncType::Existing(_)` rather than `LoweredFuncType::New(_)` when `gc` is `true`, and I couldn't figure out how to extract the core parameter list from the former.  Do you have any tips?

Reply:
Nah that's fine to leave as a panic for now, it's more the silent/subtle bugs of switching options by accident I'm worried about. Overall this is a really minor concern so if you'd prefer it's ok to leave as-is.